### PR TITLE
Homework 38 completed

### PR DESCRIPTION
**Pull request by Borisko130**

### CODE REVIEW

**Review by vladkuzya — APPROVED** [review]
Good job)

File: src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import com.dev.cinema.lib.Service;
+import com.dev.cinema.model.MovieSession;
+import com.dev.cinema.service.MovieSessionService;
+import java.time.LocalDate;
+import java.util.List;
+
+@Service
+public class MovieSessionServiceImpl implements MovieSessionService {
+    @Inject
+    MovieSessionDao movieSessionDao;
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java:14]: what about access modifier?

File: src/main/java/com/dev/cinema/service/impl/CinemaHallServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import com.dev.cinema.lib.Inject;
+import com.dev.cinema.lib.Service;
+import com.dev.cinema.model.CinemaHall;
+import com.dev.cinema.service.CinemaHallService;
+import java.util.List;
+
+@Service
+public class CinemaHallServiceImpl implements CinemaHallService {
+    @Inject
+    CinemaHallDao cinemaHallDao;
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/CinemaHallServiceImpl.java:13]: same story

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<MovieSession> query = session.createQuery("FROM MovieSession "
+                    + "WHERE movie.id = :movieId AND showTime "
+                    + "BETWEEN :timeStart AND :timeEnd", MovieSession.class);
+            query.setParameter("movieId", movieId);
+            query.setParameter("timeStart", date.atTime(0, 0, 0));
+            query.setParameter("timeEnd", date.atTime(23,59,59));
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java:46]: let's better use constant of LocalTime here

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<MovieSession> query = session.createQuery("FROM MovieSession "
+                    + "WHERE movie.id = :movieId AND showTime "
+                    + "BETWEEN :timeStart AND :timeEnd", MovieSession.class);
+            query.setParameter("movieId", movieId);
+            query.setParameter("timeStart", date.atTime(0, 0, 0));
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java:45]: use atStartOfDay()

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert MovieSession entity", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java:30]: let's make message more informative and add some info about entity we were trying to save (same goes for other methods with input parameters, where we throw exception)

File: src/main/java/com/dev/cinema/model/MovieSession.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+@Entity
+@Table(name = "movie_sessions")
+public class MovieSession {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    private Movie movie;
+    @ManyToOne
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/MovieSession.java:19]: this is for your info, but keep in mind that toOne relations have fetchType eager by default, which is not always quite good

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**